protocol/lib/ante/nested_msg.go

```diff
@@ -2,73 +2,99 @@ package ante
 
 import (
 	"fmt"
+	"strings"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	authz "github.com/cosmos/cosmos-sdk/x/authz"
 	gov "github.com/cosmos/cosmos-sdk/x/gov/types/v1"
+	"github.com/dydxprotocol/v4-chain/protocol/lib/metrics"
 )
 
+const DYDX_MSG_PREFIX = "/dydxprotocol"
+
 // IsNestedMsg returns true if the given msg is a nested msg.
 func IsNestedMsg(msg sdk.Msg) bool {
 	switch msg.(type) {
 	case
 		// ------- CosmosSDK default modules
+		// authz
+		*authz.MsgExec,
 		// gov
 		*gov.MsgSubmitProposal:
 		return true
 	}
 	return false
 }
 
+// IsDydxMsg returns true if the given msg is a dYdX custom msg.
+func IsDydxMsg(msg sdk.Msg) bool {
+	return strings.HasPrefix(sdk.MsgTypeURL(msg), DYDX_MSG_PREFIX)
+}
+
 // ValidateNestedMsg returns err if the given msg is an invalid nested msg.
 func ValidateNestedMsg(msg sdk.Msg) error {
 	if !IsNestedMsg(msg) {
 		return fmt.Errorf("not a nested msg")
 	}
 
-	// Get inner msgs.
-	innerMsgs, err := getInnerMsgs(msg)
-	if err != nil {
-		return err
-	}
-
 	// Check that the inner msgs are valid.
-	if err := validateInnerMsg(innerMsgs); err != nil {
+	if err := validateInnerMsg(msg); err != nil {
 		return err
 	}
 
 	return nil // is valid nested msg.
 }
 
-// getInnerMsgs returns the inner msgs of the given msg.
-func getInnerMsgs(msg sdk.Msg) ([]sdk.Msg, error) {
-	switch msg := msg.(type) {
-	case
-		*gov.MsgSubmitProposal:
-		return msg.GetMsgs()
-	default:
-		return nil, fmt.Errorf("unsupported msg type: %T", msg)
+// validateInnerMsg returns err if the given inner msgs contain an invalid msg.
+func validateInnerMsg(msg sdk.Msg) error {
+	// Get inner msgs.
+	innerMsgs, err := getInnerMsgs(msg)
+	if err != nil {
+		return err
 	}
-}
 
-// validateInnerMsg returns err if the given inner msgs contain an invalid msg.
-func validateInnerMsg(innerMsgs []sdk.Msg) error {
-	for _, msg := range innerMsgs {
+	for _, inner := range innerMsgs {
 		// 1. unsupported msgs.
-		if IsUnsupportedMsg(msg) {
+		if IsUnsupportedMsg(inner) {
 			return fmt.Errorf("Invalid nested msg: unsupported msg type")
 		}
 
 		// 2. app-injected msgs.
-		if IsAppInjectedMsg(msg) {
+		if IsAppInjectedMsg(inner) {
 			return fmt.Errorf("Invalid nested msg: app-injected msg type")
 		}
 
 		// 3. double-nested msgs.
-		if IsNestedMsg(msg) {
+		if IsNestedMsg(inner) {
 			return fmt.Errorf("Invalid nested msg: double-nested msg type")
 		}
 
+		// 4.Reject nested dydxprotocol messages in `MsgExec`.
+		if _, ok := msg.(*authz.MsgExec); ok {
+			metrics.IncrCountMetricWithLabels(
+				metrics.Ante,
+				metrics.MsgExec,
+				metrics.GetLabelForStringValue(metrics.InnerMsg, sdk.MsgTypeURL(inner)),
+			)
+			if IsDydxMsg(inner) {
```

Comment:
Does this mean we don't allow `MsgExec` to call any of the dYdX messages (incl. order placing, subaccount transfer)? I can see why that would make sense to limit security risk. 

Is there any context/discussion behind this? If so could you document briefly?

Reply:
ah there was a comment discussion in the authz one pager that got resolved. I added a short section in the doc saying that dydx messages will be explicitly disallowed for now and an allowlist can be maintained in the future if there are use cases.